GPT/gpt-function-calling.py

```diff
@@ -0,0 +1,229 @@
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The text to display",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            },
+            {
+                "type": "function",
+                "function": {
+                    "name": "search_for_command",
+                    "description": "search_for_command(str: string) - this searches for a command in the VSCode command palette. If I ask you to do something, please use this command to search for an appropriate command.",
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The command to search for",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            }
+        ],
+        "max_tokens": 2024,
+        "temperature": 0.6,
+        "n": 1,
+        "stop": None,
+        "model": settings.get("user.openai_model"),
+    }
+
+
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+
+    if response.status_code == 200:
+        notify("GPT Task Completed")
+        print(response.json())
+
+        try:
+            tool_calls = response.json()["choices"][0]["message"]["tool_calls"]
+            while tool_calls:
+                tool = tool_calls.pop()
+                first_argument = tool['function']['arguments']
+                try:
+                    first_argument = json.loads(first_argument)['str']
+                except Exception as e:
+                    notify(f"Argument Jason was malformed: {e}")
+                if tool['function']['name'] == 'insert':
+                    insert_response(first_argument)
+                elif tool['function']['name'] == 'display':
+                    actions.user.display_response(first_argument)
+                elif tool['function']['name'] == 'notify':
+                    actions.user.notify_user(first_argument)
+                elif tool['function']['name'] == 'search_for_command':
+                    actions.user.search_for_command(first_argument)
+        except Exception as e:
+            notify(f"No tool_calls found in response from LLM: {e}")
+
+        content = (response.json()["choices"][0]["message"]["content"] or "").strip()
+        if len(content) != 0:
+            actions.user.display_response(content)
+
+    else:
+        notify("GPT Failure: Check API Key, Model, or Prompt")
+        print(response.json())
+
+
+@mod.action_class
+class UserActions:
+    def gpt_go(utterance: str, selected_text: str) -> str:
```

Comment:
I would rename this to `gpt_call_function` or something similar that you like that more clearly specifies what it does. I think `go` is a bit unclear to new users.